File: src/dmx/dmx.js
// ============================================================================
// dmx.js — Sistema de Iluminação DMX (ArtNet + Lumikit)
// ============================================================================
// Gerencia a integração com o sistema de iluminação:
// - startDmxApp: inicia o ArtNetToDMX.exe (conversor ArtNet -> DMX via FTDI)
// - resetDmxSystem: reset completo (USB + software) com PowerShell elevado
// - updateLumikitConfig: atualiza o IP no arquivo "info" do ArtNetToDMX
// Fluxo típico: ArtNetToDMX.exe -> USB FTDI -> DMX -> LumikitSHOW.exe
// ============================================================================

const os = require('os');
const fs = require('fs');
const path = require('path');
const { exec, spawn } = require('child_process');

let ctx;

// Inicia (ou verifica) o ArtNetToDMX.exe.
// Se force=true, força kill e respawn. Se false, só abre se não estiver rodando.
function startDmxApp(force = false) {
    const exePath = path.join(ctx.rootDir, 'ArtNetToDMX_FTDI', 'ArtNetToDMX.exe');

    // --- AUTO-CONFIGURAÇÃO DE IP ---
    // Sempre garante que o arquivo info está com o IP correto para a rede atual,
    // independente de o aplicativo já estar aberto ou não.
    updateLumikitConfig();

    // Verifica se o processo já existe na lista do Windows
    exec('tasklist /FI "IMAGENAME eq ArtNetToDMX.exe"', (err, stdout) => {
        const isRunning = stdout.toLowerCase().includes('artnettodmx.exe');

        if (isRunning && !force) {
            console.log('💡 [DMX] Aplicativo de luz já está em execução. Nenhuma ação necessária no boot.');
            return;
        }

        if (isRunning && force) {
            console.log('♻️ [DMX] Forçando reinicialização do aplicativo...');
            exec('taskkill /F /IM ArtNetToDMX.exe', () => spawnDmx());
        } else {
            console.log('🎬 [DMX] Iniciando aplicativo de luz...');
            spawnDmx();
        }
    });

    function spawnDmx() {
        if (!fs.existsSync(exePath)) return console.error('❌ [DMX] Executável não encontrado em', exePath);

        try {
            const child = spawn(exePath, [], {
                cwd: path.dirname(exePath),
                detached: true,
                stdio: 'ignore'
            });
            child.unref();
            console.log('🚀 [DMX] Sistema de luz online!');
        } catch (e) {
            console.error('❌ [DMX] Erro ao abrir executável:', e.message);
        }
    }
}

// Procedimento completo de reset: mata processos, reseta USB via pnputil, reabre apps.
// Usado quando o DMX para de responder e precisa de um reset de hardware.
function resetDmxSystem() {
    console.log('🚀 [DMX] Iniciando procedimento de reset de hardware (USB) e software...');

    // 1. Matar o LumikitSHOW.exe
    exec('taskkill /F /IM LumikitSHOW.exe', () => {
        // 2. Matar o ArtNetToDMX.exe
        exec('taskkill /F /IM ArtNetToDMX.exe', () => {
            // 3. Delay para o Windows processar o fechamento
            setTimeout(() => {
                console.log('🔧 [DMX] Executando reset USB elevado via PowerShell (pnputil)...');
                // Adicionado -Wait para o exec() só terminar quando o reset físico for concluído.
                const psCommand = `powershell -Command "Start-Process powershell -ArgumentList '-NoProfile -Command $dev = Get-PnpDevice | Where-Object { $_.InstanceId -like ''*VID_0403&PID_6001*'' -or $_.FriendlyName -like ''*USB Serial Converter*'' } | Select-Object -First 1; if ($dev) { pnputil /restart-device $dev.InstanceId }' -Verb RunAs -WindowStyle Hidden -Wait"`;

                exec(psCommand, (psErr) => {
                    if (psErr) console.error('❌ [DMX] Erro ao disparar reset elevado:', psErr.message);
                    else console.log('✅ [DMX] Comando de reset enviado para o Windows e concluído.');

                    // 4. Aguarda 3s para o Windows re-enumerar o dispositivo USB
                    setTimeout(() => {
                        console.log('🎬 [DMX] Iniciando ArtNetToDMX...');
                        startDmxApp(true); // 'true' garante que ele force caso tenha sobrado algum zumbi

                        // 5. Aguarda mais 3s para o ArtNetToDMX carregar e escutar a porta 6454
                        setTimeout(() => {
                            const lumikitPath = "C:\\Program Files\\Lumikit\\LumikitSHOW.exe";
                            if (fs.existsSync(lumikitPath)) {
                                console.log('🎬 [DMX] Iniciando LumikitSHOW...');
                                try {
                                    const child = spawn(lumikitPath, [], {
                                        cwd: path.dirname(lumikitPath),
                                        detached: true,
                                        stdio: 'ignore'
                                    });
                                    child.unref();
                                } catch (e) {
                                    console.error('❌ [DMX] Erro ao abrir Lumikit:', e.message);
                                }
                            } else {
                                console.error('❌ [DMX] Executável Lumikit não encontrado em', lumikitPath);
                            }
                        }, 3000);
                    }, 3000);
                });
            }, 1000);
        });
    });
}

// Configura o IP local no arquivo "info" do ArtNetToDMX.
// Compara os IPs da lista lumikit_ips com as interfaces de rede ativas e
// escreve o IP correto no arquivo de configuração do conversor ArtNet->DMX.
function updateLumikitConfig() {
    const config = ctx.loadConfig();
    const lumikitIps = config.lumikit_ips || [];
    if (lumikitIps.length === 0) return;

    const infoPath = path.join(ctx.rootDir, 'ArtNetToDMX_FTDI', 'info');
    const interfaces = os.networkInterfaces();
    let localIps = [];
    for (const k in interfaces) {
        for (const k2 in interfaces[k]) {
            const address = interfaces[k][k2];
            if (address.family === 'IPv4' && !address.internal) {
                localIps.push(address.address);
            }
        }
    }

    const match = lumikitIps.find(ip => localIps.includes(ip));

    if (match) {
        try {
            // Se o arquivo não existe, cria um novo com as configurações padrão
            if (!fs.existsSync(infoPath)) {
                console.log(`📝 [DMX] Arquivo "info" não encontrado. Criando um novo para o IP ${match}...`);
                const defaultContent = `IP: ${match}\nUni: 0\nOneUni: true\nAutostart: true\n`;
                fs.writeFileSync(infoPath, defaultContent);
                return;
            }

            let infoContent = fs.readFileSync(infoPath, 'utf8');
            const newContent = infoContent.replace(/^IP:.*$/m, `IP: ${match}`);

            if (infoContent !== newContent) {
                fs.writeFileSync(infoPath, newContent);
                console.log(`🌐 [DMX] IP configurado automaticamente no arquivo info: ${match}`);
            } else {
                console.log(`🌐 [DMX] IP ${match} já estava configurado corretamente.`);
            }
        } catch (err) {
            console.error('❌ [DMX] Erro ao gravar/criar o arquivo info:', err.message);
        }
    } else {
        console.warn('⚠️ [DMX] Nenhum IP da lista "lumikit_ips" bate com as redes ativas deste PC.');
    }
}

function initDmx(appCtx) {
    ctx = appCtx;
    ctx.startDmxApp = startDmxApp;
    ctx.resetDmxSystem = resetDmxSystem;
    ctx.updateLumikitConfig = updateLumikitConfig;
}

module.exports = { initDmx };

File: server_rust/src/dmx.rs
use std::path::Path;
use std::process::{Command, Stdio};
use std::time;

pub fn start_dmx_app(force: bool, root_dir: &str) {
    let exe_path_str = format!("{}\\ArtNetToDMX_FTDI\\ArtNetToDMX.exe", root_dir);
    let exe_path = Path::new(&exe_path_str);

    update_lumikit_config(root_dir);

    // Simple check via tasklist
    let output = Command::new("tasklist")
        .arg("/FI")
        .arg("IMAGENAME eq ArtNetToDMX.exe")
        .output()
        .expect("Failed to execute tasklist");

    let stdout = String::from_utf8_lossy(&output.stdout).to_lowercase();
    let is_running = stdout.contains("artnettodmx.exe");

    if is_running && !force {
        println!("Y' [DMX] Aplicativo de luz j est em execuo. Nenhuma ao necessria no boot.");
        return;
    }

    if is_running && force {
        println!("T? [DMX] Forando reinicializao do aplicativo...");
        let _ = Command::new("taskkill")
            .args(&["/F", "/IM", "ArtNetToDMX.exe"])
            .output();
        spawn_dmx(exe_path);
    } else {
        println!("YZ [DMX] Iniciando aplicativo de luz...");
        spawn_dmx(exe_path);
    }
}

fn spawn_dmx(exe_path: &Path) {
    if !exe_path.exists() {
        eprintln!("?O [DMX] Executvel no encontrado em {:?}", exe_path);
        return;
    }

    if let Some(parent) = exe_path.parent() {
        match Command::new(exe_path)
            .current_dir(parent)
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
        {
            Ok(_) => println!("Ys? [DMX] Sistema de luz online!"),
            Err(e) => eprintln!("?O [DMX] Erro ao abrir executvel: {}", e),
        }
    }
}

pub fn reset_dmx_system(root_dir: String) {
    println!("Ys? [DMX] Iniciando procedimento de reset de hardware (USB) e software...");

    tokio::spawn(async move {
        let _ = Command::new("taskkill")
            .args(&["/F", "/IM", "LumikitSHOW.exe"])
            .output();
        let _ = Command::new("taskkill")
            .args(&["/F", "/IM", "ArtNetToDMX.exe"])
            .output();

        tokio::time::sleep(time::Duration::from_millis(1000)).await;

        println!("Y\" [DMX] Executando reset USB elevado via PowerShell (pnputil)...");
        let ps_cmd = "Start-Process powershell -ArgumentList '-NoProfile -Command  = Get-PnpDevice | Where-Object { .InstanceId -like ''*VID_0403&PID_6001*'' -or .FriendlyName -like ''*USB Serial Converter*'' } | Select-Object -First 1; if () { pnputil /restart-device .InstanceId }' -Verb RunAs -WindowStyle Hidden -Wait";

        match Command::new("powershell")
            .args(&["-Command", ps_cmd])
            .output()
        {
            Ok(_) => println!("o. [DMX] Comando de reset enviado para o Windows e concludo."),
            Err(e) => eprintln!("?O [DMX] Erro ao disparar reset elevado: {}", e),
        }

        tokio::time::sleep(time::Duration::from_millis(3000)).await;
        println!("YZ [DMX] Iniciando ArtNetToDMX...");
        start_dmx_app(true, &root_dir);

        tokio::time::sleep(time::Duration::from_millis(3000)).await;
        let lumikit_path = Path::new("C:\\Program Files\\Lumikit\\LumikitSHOW.exe");
        if lumikit_path.exists() {
            println!("YZ [DMX] Iniciando LumikitSHOW...");
            let _ = Command::new(lumikit_path)
                .current_dir(lumikit_path.parent().unwrap())
                .stdout(Stdio::null())
                .stderr(Stdio::null())
                .spawn();
        } else {
            eprintln!(
                "?O [DMX] Executvel Lumikit no encontrado em {:?}",
                lumikit_path
            );
        }
    });
}

pub fn update_lumikit_config(root_dir: &str) {
    let config = crate::config::AppConfig::load();
    let lumikit_ips = config.lumikit_ips;
    if lumikit_ips.is_empty() {
        return;
    }

    let mut local_ips = Vec::new();
    #[cfg(target_os = "windows")]
    {
        if let Ok(output) = Command::new("ipconfig").output() {
            let stdout = String::from_utf8_lossy(&output.stdout);
            for line in stdout.lines() {
                if line.contains("IPv4") {
                    if let Some(ip_part) = line.split(':').last() {
                        let ip = ip_part.trim();
                        if !ip.is_empty() {
                            local_ips.push(ip.to_string());
                        }
                    }
                }
            }
        }
    }

    let found_match = lumikit_ips.iter().find(|ip| local_ips.contains(ip));

    if let Some(matched_ip) = found_match {
        let info_path = Path::new(root_dir).join("ArtNetToDMX_FTDI").join("info");

        if !info_path.exists() {
            println!(
                "📝 [DMX] Arquivo \"info\" nao encontrado. Criando um novo para o IP {}...",
                matched_ip
            );
            let default_content = format!(
                "IP: {}\nUni: 0\nOneUni: true\nAutostart: true\n",
                matched_ip
            );
            if let Err(e) = std::fs::write(&info_path, default_content) {
                eprintln!("❌ [DMX] Erro ao criar o arquivo info: {}", e);
            }
            return;
        }

        if let Ok(content) = std::fs::read_to_string(&info_path) {
            let lines: Vec<String> = content
                .lines()
                .map(|line| {
                    if line.starts_with("IP:") {
                        format!("IP: {}", matched_ip)
                    } else {
                        line.to_string()
                    }
                })
                .collect();
            let new_content = lines.join("\n") + "\n";

            if content != new_content {
                if let Err(e) = std::fs::write(&info_path, new_content) {
                    eprintln!("❌ [DMX] Erro ao atualizar o arquivo info: {}", e);
                } else {
                    println!(
                        "🌐 [DMX] IP configurado automaticamente no arquivo info: {}",
                        matched_ip
                    );
                }
            } else {
                println!(
                    "🌐 [DMX] IP {} ja estava configurado corretamente.",
                    matched_ip
                );
            }
        }
    } else {
        eprintln!("⚠️ [DMX] Nenhum IP da lista \"lumikit_ips\" bate com as redes ativas deste PC.");
    }
}
